Language test improvements (#7954)
NixOS/nix: doc/manual/src/contributing/testing.md

> # Running tests

Reviewer: I encountered the same thing just yesterday. I'd quickly merge moving that section out into its own page if you split out a PR for that.

$ ./mk/run-test.sh tests/${testName}.sh
ran test tests/${testName}.sh... [PASS]
```

To see the complete output, one can also run:

```shell-session
$ ./mk/debug-test.sh tests/${testName}.sh
+ foo
output from foo
+ bar
output from bar
...
```

The test script will then be traced with `set -x` and the output displayed as it happens, regardless of whether the test succeeds or fails.

### Debugging failing functional tests

When a functional test fails, it usually does so somewhere in the middle of the script.

To figure out what's wrong, it is convenient to run the test regularly up to the failing `nix` command, and then run that command with a debugger like GDB.

For example, if the script looks like:

```bash
foo
nix blah blub
bar
```
edit it like so:

```diff
 foo
-nix blah blub
+gdb --args nix blah blub
 bar
```

Then, running the test with `./mk/debug-test.sh` will drop you into GDB once the script reaches that point:

```shell-session
$ ./mk/debug-test.sh tests/${testName}.sh
...
+ gdb blash blub
GNU gdb (GDB) 12.1
...
(gdb)
```

One can debug the Nix invocation in all the usual ways.
For example, enter `run` to start the Nix invocation.

### Characterization testing

> Occasionally, Nix utilizes a technique called [Characterization Testing](doc/manual/src/contributing/hacking.md) as part of the functional tests.

Reviewer: That link seems wrong

One can debug the Nix invocation in all the usual ways.
For example, enter `run` to start the Nix invocation.

### Characterization testing

Occasionally, Nix utilizes a technique called [Characterization Testing](https://en.wikipedia.org/wiki/Characterization_test) as part of the functional tests.
This technique is to include the exact output/behavior of a former version of Nix in a test in order to check that Nix continues to produce the same behavior going forward.

> For example, this technique is used for the language tests (`test/lang.sh`).

Author: What to link? If we link any specific branch it could fall out of sync.

> For example, this technique is used for the language tests (`test/lang.sh`).

Reviewer 2: You could use a Github permalink, like https://github.com/NixOS/nix/blob/45a3e497da20e338b747c1e0ca1a1af5531080df/tests/lang.sh. Just make sure the same hash makes into the repo mainline.

> For example, this technique is used for the language tests (`test/lang.sh`).

Reviewer 3: I think we should do the simple thing and fix the tooling if it doesn't render correctly, or not link at all. `lang.sh` *isn't* documentation, and most of the time you'll be _typing_ it in your editor / IDE / terminal anyway.

```suggestion
For example, this technique is used for the language tests ([`test/lang.sh`](../../../../test/lang.sh)).
```

> For example, this technique is used for the language tests (`test/lang.sh`).

Author: Yeah I want to document just one version, and if we can't make this go I rather just drop the reference to the file at all and just say "the language tests".

> For example, this technique is used for the language tests (`test/lang.sh`).

Reviewer: Contributing stuff can point to GitHub master. Other than that I agree we should fix the tooling and make it render correctly by default.

It is frequently useful to regenerate the expected output.
To do that, rerun the failed test with `_NIX_TEST_ACCEPT=1`.
(At least, this is the convention we've used for `tests/lang.sh`.
If we add more characterization testing we should always strive to be consistent.)

An interesting situation to document is the case when these tests are "overfitted".
For example, the language tests store both the expected result in the case of success, and expected diagnostics (errors and warnings).
The expected successful output is supposed to be highly stable --- we do not intend to make breaking changes to the (stable parts of) the Nix language.

> However, the errors and warnings are not stable in this way, we are free to change how they are displayed at any time.

Reviewer: ```suggestion
However, the errors and warnings are not stable in this way.
We are free to change how they are displayed at any time.
```

> Why bother to track these diagnostics if they are not stable then?
> The answer is that just because we Nix developers are free to change them whenever we like doesn't mean we don't care about them.
> On the contrary, errors and warnings are incredibly important to users.
> By recording the expected output, the test suite guards against accidental changes, and ensure the *result* (not just the code that implements it) of the diagnostic code paths are code review.
> Regressions are caught, and improvements always show up in code review.

Reviewer: ```suggestion
While diagnostic outputs are not a stable interface, they are important to users.
By recording the expected output, the test suite guards against accidental changes, and ensure the *result* (not just the code that implements it) of the diagnostic code paths are under code review.
Regressions are caught, and improvements always show up in code review.
```

Author: I am worried that the new concise version rapidly goes over central point.

This rhetorical question was basically what @edolstra was asking, and I think it is a good question: anything that is not unstable is not normative behavior, so why the hell has we testing it? (emphasizing "not normative" is probably better than "not stable", I am realizing now.)

If you think the rhetorical question is too chatty, we can pick a different form, like

> It may be surprising that we would test non-normative behavior like diagnostic outputs.
> After all, the point of tests is usually to pin down just normative.
> If tests pinned down all behavior, normative and non-normative alike, they would just be a second copy of the implementation!

(but perhaps shorter!)

The point is not to write a question per se, but to get the readers attention --- whether the reader is already surprised or not. This is the key point, and we don't want them to miss it.

Author: On a meta point, while one might argue the main portion of the manual is strictly reference not explanation, I think the contributing chapter is free to cover all quadrants, as it currently does, as `nix.dev` etc. will not be covering contributing to Nix.

Reviewer: > It may be surprising that we would test non-normative behavior like diagnostic outputs.

This one's good for prepending it to my suggestion. I'm in general against being chatty in documentation.

Author: ```suggestion
It may be surprising that we would test non-normative behavior like diagnostic outputs.
While diagnostic outputs are not a stable interface, they are important to users.
By recording the expected output, the test suite guards against accidental changes, and ensure the *result* (not just the code that implements it) of the diagnostic code paths are under code review.
Regressions are caught, and improvements always show up in code review.
```
OK!

> These are good benefits, but its important that they are not offset by newly imposed costs.
> To that end, it much be very easy to regenerate the tests so characterization testing doesn't make it harder to intentionally change these unstable interfaces.
> That is why having an easy way to regenerate the expected output, e.g. with `_NIX_TEST_ACCEPT=1`, is very important.

Reviewer: ```suggestion
To ensure that characterization testing doesn't make it harder to intentionally change these interfaces, there always must be an easy way to regenerate the expected output, e.g. with `_NIX_TEST_ACCEPT=1`.
```

Reviewer 3: ```suggestion
To ensure that characterization testing doesn't make it harder to intentionally change these interfaces, there always must be an easy way to regenerate the expected output, as we do with `_NIX_TEST_ACCEPT=1`.
```
